File: src/scripts/prove_pack_extractable.py
#!/usr/bin/env python3
"""Prove a pack + its declared dependency closure is self-contained.

road-to-6.0.0-D Phase 1 Step 6 — the *extraction proof*. It de-risks the
monorepo collapse (Phase 3) by proving that a pack moved into the flat
``src/`` library can still be lifted back out into a standalone package:
every artefact the pack references must live inside the pack's own
``requires`` closure (plus the always-available foundation packs). A
reference that points OUTSIDE the closure is a dangling edge — the slice
would not build standalone, so re-split is no longer possible.

What "its own tests pass" means for a markdown artefact library: the
isolated slice has (1) zero dangling skill/rule references in frontmatter,
(2) zero dangling markdown links into the artefact library, and (3) a
``pack.yaml`` whose ``requires`` graph stays acyclic within the closure.
Those are the structural invariants the per-pack CI gates assert; proving
them on the isolated closure is the standalone-build proof.

Usage:
  python3 scripts/prove_pack_extractable.py <pack-id> [--json]

Exit codes: 0 = extractable · 1 = dangling reference(s) · 3 = unknown pack.
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

import yaml

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _lib import agent_src  # noqa: E402

ROOT = Path(__file__).resolve().parents[2]
PACKS_VOCAB = ROOT / "config" / "discovery" / "packs.yml"
# Always-available foundation packs: every pack may reference these without
# declaring them in `requires` (the 6.0.0-D council's resolution of the
# ambiguous "core" in the boundary rule — engineering-base + meta are the
# implicit foundation, alongside the legacy physical `core`).
FOUNDATION = {"core", "engineering-base", "meta"}
_LINK_RE = re.compile(r"\]\(([^)#?]+\.md)(?:[#?][^)]*)?\)")


def _frontmatter(path: Path) -> dict:
    text = path.read_text("utf-8", errors="replace")
    if not text.startswith("---"):
        return {}
    end = text.find("\n---", 4)
    if end == -1:
        return {}
    try:
        data = yaml.safe_load(text[4:end])
    except yaml.YAMLError:
        return {}
    return data if isinstance(data, dict) else {}


def _pack_requires() -> dict[str, set[str]]:
    """Map pack id -> direct `requires` set, from the discovery vocab."""
    graph: dict[str, set[str]] = {}
    vocab = yaml.safe_load(PACKS_VOCAB.read_text("utf-8")) or []
    for entry in vocab:
        if isinstance(entry, dict) and isinstance(entry.get("id"), str):
            graph[entry["id"]] = set(entry.get("requires") or [])
    return graph


def _closure(pack: str, graph: dict[str, set[str]]) -> set[str]:
    """Transitive `requires` closure of ``pack`` (inclusive) + foundation."""
    seen: set[str] = set()
    stack = [pack]
    while stack:
        p = stack.pop()
        if p in seen:
            continue
        seen.add(p)
        stack.extend(graph.get(p, set()) - seen)
    return seen | FOUNDATION


def _library_index() -> tuple[dict[str, set[str]], dict[str, set[str]]]:
    """Build (slug_to_packs, logicalpath_to_packs) over the flat library."""
    slug_packs: dict[str, set[str]] = {}
    path_packs: dict[str, set[str]] = {}
    for phys, logical in agent_src.iter_all_sources():
        if not logical.endswith(".md"):
            continue
        fm = _frontmatter(phys)
        packs = set(p for p in (fm.get("packs") or []) if isinstance(p, str))
        if not packs:
            continue
        path_packs[logical] = packs
        if logical.startswith("skills/") and logical.endswith("/SKILL.md"):
            slug_packs[logical.split("/")[1]] = packs
        elif logical.startswith("rules/"):
            slug_packs[logical[len("rules/"):-len(".md")]] = packs
    return slug_packs, path_packs


def prove(pack: str) -> tuple[bool, list[str], list[str], set[str]]:
    """Return (extractable, hard_dangling, advisory_warnings, closure).

    The dependency axis is frontmatter (6.0.0-D council convergence): a
    HARD edge is a ``skills:`` / ``rules:`` include — it must resolve inside
    the closure or the standalone slice fails to build. A markdown link into
    another pack is an ADVISORY cross-reference (a "route to" / "see also")
    — it is reported as a warning but does NOT block extraction, because the
    skill still functions when the linked alternative is simply not installed.
    """
    graph = _pack_requires()
    if pack not in graph:
        return False, [f"unknown pack id '{pack}' (not in {PACKS_VOCAB})"], [], set()
    closure = _closure(pack, graph)
    slug_packs, path_packs = _library_index()

    def in_closure(packs: set[str]) -> bool:
        return bool(packs & closure)

    hard: list[str] = []
    advisory: list[str] = []
    members = [(logical, phys) for phys, logical in agent_src.iter_all_sources()
               if logical.endswith(".md") and pack in path_packs.get(logical, set())]
    for logical, phys in members:
        fm = _frontmatter(phys)
        # (1) HARD: frontmatter skill/rule includes — the build-dependency axis
        for slug in (fm.get("skills") or []) + (fm.get("rules") or []):
            if not isinstance(slug, str):
                continue
            target = slug_packs.get(slug)
            if target is None:
                continue  # command-routing target / non-library slug — not a pack edge
            if not in_closure(target):
                hard.append(
                    f"{logical}: include '{slug}' lives in {sorted(target)} — outside closure")
        # (2) ADVISORY: markdown links into the artefact library
        for raw in _LINK_RE.findall(phys.read_text("utf-8", errors="replace")):
            tgt = agent_src.strip_source_prefix(raw) or _resolve_rel(phys, raw)
            if tgt is None or tgt not in path_packs:
                continue
            if not in_closure(path_packs[tgt]):
                advisory.append(
                    f"{logical}: advisory link → {tgt} ({sorted(path_packs[tgt])}) — "
                    f"not installed when '{pack}' is extracted alone")
    return (not hard), hard, advisory, closure


def _resolve_rel(source: Path, raw: str) -> str | None:
    try:
        resolved = (source.parent / raw).resolve()
        rel = resolved.relative_to(ROOT).as_posix()
    except (OSError, ValueError):
        return None
    return agent_src.strip_source_prefix(rel)


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("pack", help="pack id (e.g. laravel)")
    ap.add_argument("--json", action="store_true")
    args = ap.parse_args()
    ok, hard, advisory, closure = prove(args.pack)
    if not closure:
        print(f"❌  {hard[0]}", file=sys.stderr)
        return 3
    if args.json:
        json.dump({"pack": args.pack, "extractable": ok,
                   "closure": sorted(closure), "hard_dangling": hard,
                   "advisory": advisory}, sys.stdout, indent=2)
        sys.stdout.write("\n")
    elif ok:
        print(f"✅  '{args.pack}' is extractable — closure {sorted(closure)}, "
              f"0 hard dangling references.")
        for a in advisory:
            print(f"   ⚠️  {a}")
    else:
        for d in hard:
            print(f"❌  {d}", file=sys.stderr)
        print(f"\n{len(hard)} hard dangling reference(s) — '{args.pack}' is NOT "
              f"standalone-extractable.", file=sys.stderr)
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/prove_pack_extractable.ts
#!/usr/bin/env node
/**
 * Prove a pack + its declared dependency closure is self-contained.
 *
 * TypeScript twin of `prove_pack_extractable.py` (Phase 8 / Wave 8g).
 *
 * road-to-6.0.0-D Phase 1 Step 6 — the *extraction proof*. It de-risks the
 * monorepo collapse (Phase 3) by proving that a pack moved into the flat
 * `src/` library can still be lifted back out into a standalone package:
 * every artefact the pack references must live inside the pack's own
 * `requires` closure (plus the always-available foundation packs). A
 * reference that points OUTSIDE the closure is a dangling edge — the slice
 * would not build standalone, so re-split is no longer possible.
 *
 * What "its own tests pass" means for a markdown artefact library: the
 * isolated slice has (1) zero dangling skill/rule references in frontmatter,
 * (2) zero dangling markdown links into the artefact library, and (3) a
 * `pack.yaml` whose `requires` graph stays acyclic within the closure.
 * Those are the structural invariants the per-pack CI gates assert; proving
 * them on the isolated closure is the standalone-build proof.
 *
 * Usage:
 *   node prove_pack_extractable.js <pack-id> [--json]
 *
 * Exit codes: 0 = extractable · 1 = dangling reference(s) · 3 = unknown pack.
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

import { parse as parseYaml } from 'yaml';

import * as agent_src from './_lib/agent_src.js';

const _HERE = fileURLToPath(import.meta.url);

const ROOT = agent_src.ROOT();
const PACKS_VOCAB = path.join(ROOT, 'src', 'config', 'discovery', 'packs.yml');
// Always-available foundation packs: every pack may reference these without
// declaring them in `requires` (the 6.0.0-D council's resolution of the
// ambiguous "core" in the boundary rule — engineering-base + meta are the
// implicit foundation, alongside the legacy physical `core`).
const FOUNDATION = new Set<string>(['core', 'engineering-base', 'meta']);
const _LINK_RE = /\]\(([^)#?]+\.md)(?:[#?][^)]*)?\)/g;

function _frontmatter(p: string): Record<string, unknown> {
    let text: string;
    try {
        text = fs.readFileSync(p, 'utf-8');
    } catch {
        // Python read_text(errors="replace") never raises on decode; a missing
        // file would raise, but callers only pass existing paths.
        return {};
    }
    if (!text.startsWith('---')) {
        return {};
    }
    const end = text.indexOf('\n---', 4);
    if (end === -1) {
        return {};
    }
    let data: unknown;
    try {
        data = parseYaml(text.slice(4, end), { version: '1.1' });
    } catch {
        return {};
    }
    return data !== null && typeof data === 'object' && !Array.isArray(data)
        ? (data as Record<string, unknown>)
        : {};
}

/** Map pack id -> direct `requires` set, from the discovery vocab. */
function _pack_requires(): Map<string, Set<string>> {
    const graph = new Map<string, Set<string>>();
    const vocab = (parseYaml(fs.readFileSync(PACKS_VOCAB, 'utf-8'), { version: '1.1' }) ??
        []) as unknown;
    const entries = Array.isArray(vocab) ? vocab : [];
    for (const entry of entries) {
        if (
            entry !== null &&
            typeof entry === 'object' &&
            !Array.isArray(entry) &&
            typeof (entry as Record<string, unknown>)['id'] === 'string'
        ) {
            const e = entry as Record<string, unknown>;
            const requires = e['requires'];
            const reqList = Array.isArray(requires) ? (requires as unknown[]) : [];
            graph.set(e['id'] as string, new Set(reqList.map((x) => String(x))));
        }
    }
    return graph;
}

/** Transitive `requires` closure of `pack` (inclusive) + foundation. */
function _closure(pack: string, graph: Map<string, Set<string>>): Set<string> {
    const seen = new Set<string>();
    const stack: string[] = [pack];
    while (stack.length > 0) {
        const p = stack.pop() as string;
        if (seen.has(p)) {
            continue;
        }
        seen.add(p);
        for (const dep of graph.get(p) ?? new Set<string>()) {
            if (!seen.has(dep)) {
                stack.push(dep);
            }
        }
    }
    return new Set<string>([...seen, ...FOUNDATION]);
}

/** Build [slug_to_packs, logicalpath_to_packs] over the flat library. */
function _library_index(): [Map<string, Set<string>>, Map<string, Set<string>>] {
    const slugPacks = new Map<string, Set<string>>();
    const pathPacks = new Map<string, Set<string>>();
    for (const [phys, logical] of agent_src.iter_all_sources()) {
        if (!logical.endsWith('.md')) {
            continue;
        }
        const fm = _frontmatter(phys);
        const rawPacks = fm['packs'];
        const packsList = Array.isArray(rawPacks) ? (rawPacks as unknown[]) : [];
        const packs = new Set<string>(
            packsList.filter((x): x is string => typeof x === 'string'),
        );
        if (packs.size === 0) {
            continue;
        }
        pathPacks.set(logical, packs);
        if (logical.startsWith('skills/') && logical.endsWith('/SKILL.md')) {
            slugPacks.set(logical.split('/')[1] as string, packs);
        } else if (logical.startsWith('rules/')) {
            slugPacks.set(logical.slice('rules/'.length, logical.length - '.md'.length), packs);
        }
    }
    return [slugPacks, pathPacks];
}

export interface ProveResult {
    extractable: boolean;
    hard: string[];
    advisory: string[];
    closure: Set<string>;
}

/**
 * Return {extractable, hard, advisory, closure}.
 *
 * The dependency axis is frontmatter (6.0.0-D council convergence): a HARD
 * edge is a `skills:` / `rules:` include — it must resolve inside the
 * closure or the standalone slice fails to build. A markdown link into
 * another pack is an ADVISORY cross-reference (a "route to" / "see also") —
 * it is reported as a warning but does NOT block extraction, because the
 * skill still functions when the linked alternative is simply not installed.
 */
export function prove(pack: string): ProveResult {
    const graph = _pack_requires();
    if (!graph.has(pack)) {
        return {
            extractable: false,
            hard: [`unknown pack id '${pack}' (not in ${PACKS_VOCAB})`],
            advisory: [],
            closure: new Set<string>(),
        };
    }
    const closure = _closure(pack, graph);
    const [slugPacks, pathPacks] = _library_index();

    const inClosure = (packs: Set<string>): boolean => {
        for (const p of packs) {
            if (closure.has(p)) {
                return true;
            }
        }
        return false;
    };

    const hard: string[] = [];
    const advisory: string[] = [];
    const members: Array<[string, string]> = [];
    for (const [phys, logical] of agent_src.iter_all_sources()) {
        if (logical.endsWith('.md') && (pathPacks.get(logical) ?? new Set<string>()).has(pack)) {
            members.push([logical, phys]);
        }
    }
    for (const [logical, phys] of members) {
        const fm = _frontmatter(phys);
        // (1) HARD: frontmatter skill/rule includes — the build-dependency axis
        const rawSkills = Array.isArray(fm['skills']) ? (fm['skills'] as unknown[]) : [];
        const rawRules = Array.isArray(fm['rules']) ? (fm['rules'] as unknown[]) : [];
        for (const slug of [...rawSkills, ...rawRules]) {
            if (typeof slug !== 'string') {
                continue;
            }
            const target = slugPacks.get(slug);
            if (target === undefined) {
                continue; // command-routing target / non-library slug — not a pack edge
            }
            if (!inClosure(target)) {
                hard.push(
                    `${logical}: include '${slug}' lives in ${_sortedRepr(target)} — outside closure`,
                );
            }
        }
        // (2) ADVISORY: markdown links into the artefact library
        const physText = fs.readFileSync(phys, 'utf-8');
        for (const m of physText.matchAll(_LINK_RE)) {
            const raw = m[1] as string;
            // Python `strip_source_prefix(raw) or _resolve_rel(...)` — falls
            // through on None AND on an empty-string strip result.
            const stripped = agent_src.strip_source_prefix(raw);
            const tgt = stripped ? stripped : _resolve_rel(phys, raw);
            if (tgt === null || !pathPacks.has(tgt)) {
                continue;
            }
            if (!inClosure(pathPacks.get(tgt)!)) {
                advisory.push(
                    `${logical}: advisory link → ${tgt} (${_sortedRepr(pathPacks.get(tgt)!)}) — ` +
                        `not installed when '${pack}' is extracted alone`,
                );
            }
        }
    }
    return { extractable: hard.length === 0, hard, advisory, closure };
}

/** Mirror Python `sorted(set)` -> `repr(list)` formatting: ['a', 'b']. */
function _sortedRepr(items: Set<string>): string {
    const sorted = [...items].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
    return '[' + sorted.map((s) => `'${s}'`).join(', ') + ']';
}

function _resolve_rel(source: string, raw: string): string | null {
    let rel: string;
    try {
        const resolved = fs.realpathSync(path.resolve(path.dirname(source), raw));
        const relPath = path.relative(ROOT, resolved);
        if (relPath.startsWith('..') || path.isAbsolute(relPath)) {
            return null; // not under ROOT — mirrors relative_to ValueError
        }
        rel = relPath.split(path.sep).join('/');
    } catch {
        return null;
    }
    return agent_src.strip_source_prefix(rel);
}

/**
 * Mirror `json.dumps(obj, indent=2)` (ensure_ascii default True) for the
 * limited value shapes this script emits.
 */
function _pyJsonDumpsIndent2(obj: unknown): string {
    return _dumpValue(obj, 0);
}

function _dumpValue(value: unknown, depth: number): string {
    const pad = '  '.repeat(depth);
    const padInner = '  '.repeat(depth + 1);
    if (value === null || value === undefined) return 'null';
    if (typeof value === 'number') return String(value);
    if (typeof value === 'boolean') return value ? 'true' : 'false';
    if (typeof value === 'string') return _dumpString(value);
    if (Array.isArray(value)) {
        if (value.length === 0) return '[]';
        const items = value.map((v) => padInner + _dumpValue(v, depth + 1));
        return '[\n' + items.join(',\n') + '\n' + pad + ']';
    }
    const entries = Object.entries(value as Record<string, unknown>);
    if (entries.length === 0) return '{}';
    const items = entries.map(
        ([k, v]) => padInner + _dumpString(k) + ': ' + _dumpValue(v, depth + 1),
    );
    return '{\n' + items.join(',\n') + '\n' + pad + '}';
}

function _dumpString(s: string): string {
    let out = '"';
    for (const ch of s) {
        const code = ch.codePointAt(0) as number;
        if (ch === '"') out += '\\"';
        else if (ch === '\\') out += '\\\\';
        else if (ch === '\n') out += '\\n';
        else if (ch === '\r') out += '\\r';
        else if (ch === '\t') out += '\\t';
        else if (ch === '\b') out += '\\b';
        else if (ch === '\f') out += '\\f';
        else if (code < 0x20) out += '\\u' + code.toString(16).padStart(4, '0');
        else if (code < 0x7f) out += ch;
        else if (code <= 0xffff) out += '\\u' + code.toString(16).padStart(4, '0');
        else {
            const v = code - 0x10000;
            const hi = 0xd800 + (v >> 10);
            const lo = 0xdc00 + (v & 0x3ff);
            out += '\\u' + hi.toString(16).padStart(4, '0');
            out += '\\u' + lo.toString(16).padStart(4, '0');
        }
    }
    return out + '"';
}

class ArgError extends Error {}

interface ParsedArgs {
    pack: string;
    json: boolean;
}

function _parseArgs(argv: string[]): ParsedArgs {
    let pack: string | null = null;
    let json = false;
    for (const a of argv) {
        if (a === '--json') {
            json = true;
        } else if (a.startsWith('-') && a !== '-') {
            throw new ArgError(`unrecognized arguments: ${a}`);
        } else if (pack === null) {
            pack = a;
        } else {
            throw new ArgError(`unrecognized arguments: ${a}`);
        }
    }
    if (pack === null) {
        throw new ArgError('the following arguments are required: pack');
    }
    return { pack, json };
}

export function main(argv: string[] | null = null): number {
    let args: ParsedArgs;
    try {
        args = _parseArgs(argv ?? process.argv.slice(2));
    } catch (e) {
        if (e instanceof ArgError) {
            process.stderr.write(`prove_pack_extractable: ${e.message}\n`);
            return 2;
        }
        throw e;
    }
    const { extractable: ok, hard, advisory, closure } = prove(args.pack);
    if (closure.size === 0) {
        process.stderr.write(`❌  ${hard[0]}\n`);
        return 3;
    }
    const sortedClosure = [...closure].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
    if (args.json) {
        process.stdout.write(
            _pyJsonDumpsIndent2({
                pack: args.pack,
                extractable: ok,
                closure: sortedClosure,
                hard_dangling: hard,
                advisory,
            }),
        );
        process.stdout.write('\n');
    } else if (ok) {
        process.stdout.write(
            `✅  '${args.pack}' is extractable — closure ${_sortedReprList(sortedClosure)}, ` +
                '0 hard dangling references.\n',
        );
        for (const a of advisory) {
            process.stdout.write(`   ⚠️  ${a}\n`);
        }
    } else {
        for (const d of hard) {
            process.stderr.write(`❌  ${d}\n`);
        }
        process.stderr.write(
            `\n${hard.length} hard dangling reference(s) — '${args.pack}' is NOT ` +
                'standalone-extractable.\n',
        );
    }
    return ok ? 0 : 1;
}

/** repr() of an already-sorted list of strings. */
function _sortedReprList(items: string[]): string {
    return '[' + items.map((s) => `'${s}'`).join(', ') + ']';
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exitCode = main();
}
